pkg/server/plugin/bundlepublisher/awss3/awss3.go

```diff
@@ -0,0 +1,227 @@
+package awss3
+
+import (
+	"bytes"
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire-plugin-sdk/pluginsdk/support/bundleformat"
+	bundlepublisherv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/server/bundlepublisher/v1"
+	"github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/types"
+	configv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/service/common/config/v1"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	pluginName = "aws_s3"
+)
+
+type pluginHooks struct {
+	newS3ClientFunc func(c aws.Config) (simpleStorageService, error)
+}
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func New() *Plugin {
+	return newPlugin(newS3Client)
+}
+
+// Config holds the configuration of the plugin.
+type Config struct {
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	Region          string `hcl:"region" json:"region"`
+	Bucket          string `hcl:"bucket" json:"bucket"`
+	ObjectKey       string `hcl:"object_key" json:"object_key"`
+	Format          string `hcl:"format" json:"format"`
+}
+
+// Plugin is the main representation of this bundle publisher plugin.
+type Plugin struct {
+	bundlepublisherv1.UnsafeBundlePublisherServer
+	configv1.UnsafeConfigServer
+
+	config    *Config
+	configMtx sync.RWMutex
```

Comment:
Please excuse my ignorance, but how is the config updated dynamically in a way that would require a `RWLock`?

Reply:
The configuration can be updated at any time by a different thread if the Configure function is called.